File: code/github-actions/java/src/main/java/com/example/slugs/Slug.java
package com.example.slugs;

import java.text.Normalizer;

public final class Slug {
    private Slug() {
    }

    // "Hello, Wörld!" -> "hello-world"
    public static String from(String text) {
        var decomposed = Normalizer.normalize(text, Normalizer.Form.NFD);
        var builder = new StringBuilder(decomposed.length());
        var pendingDash = false;
        for (var i = 0; i < decomposed.length(); i++) {
            var c = decomposed.charAt(i);
            if (Character.getType(c) == Character.NON_SPACING_MARK) {
                continue;
            }
            if ((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9')) {
                if (pendingDash && !builder.isEmpty()) {
                    builder.append('-');
                }
                builder.append(Character.toLowerCase(c));
                pendingDash = false;
            } else {
                pendingDash = true;
            }
        }
        return builder.toString();
    }
}

File: code/github-actions/dotnet/Slugs/Slug.cs
using System.Globalization;
using System.Text;

namespace Slugs;

public static class Slug
{
    // "Hello, Wörld!" -> "hello-world"
    public static string From(string text)
    {
        var decomposed = text.Normalize(NormalizationForm.FormD);
        var builder = new StringBuilder(decomposed.Length);
        var pendingDash = false;
        foreach (var c in decomposed)
        {
            if (CharUnicodeInfo.GetUnicodeCategory(c) == UnicodeCategory.NonSpacingMark)
                continue;
            if (char.IsAsciiLetterOrDigit(c))
            {
                if (pendingDash && builder.Length > 0)
                    builder.Append('-');
                builder.Append(char.ToLowerInvariant(c));
                pendingDash = false;
            }
            else
            {
                pendingDash = true;
            }
        }
        return builder.ToString();
    }
}